### PR TITLE
feat(auth): user-sign-up

### LINKED ISSUE #8: [FEAT]: Authentication - User Sign Up
### **Description**
Develop and integrate a seamless user sign up feature to enhance user engagement and onboarding experience. This feature should include user registration, account creation, and authentication processes.

### **Acceptance Criteria**
 

- [x] Email and password, firstname and lastname fields should be required

 

- [x] Password validation should include a minimum of 8 alphanumeric characters

 

- [x] Password should be hashed before being stored in the database

 

- [x] A 422 error response code is returned on failed validations

 

- [x] 201 status code is returned after successful signup

 

- [x] JWT token should be generated for the user upon successful sign up and returned as part of the response payload

### **Expected Outcome**

- The user input data should be successfully sent to the backend, or the user will receive an error message if unsuccessful

### **Requirements**
**Corresponding database fields for:**

- First Name

- Last Name

- Email Address

- Password

```
Table users {
  id integer [primary key]
  identifier uuid
  first_name varchar
  last_name varchar
  email varchar
  password varchar
  created_at timestamp
  updated_at timestamp
}

```
**Hash user password**

- Develop validation logic for user-submitted data.

- Validate username/email format (e.g., email address must be a valid format).

- Ensure password meets complexity requirements (e.g., minimum length, character types).

- Implement confirmation password field validation to ensure passwords match.

- Display clear error messages for invalid inputs.

### User Registration [POST]  `/api/auth/register` 

1. Registration Endpoint

- Given a request with valid user details (ie `firstName`, `lastName`, `email`, `password`), when the user registers, then the system should create a new user account with a 201 created status code.

2. Unique Email

- Given a email that already exists, when the user tries to register, then the system should return a 400 bad request error status with an appropriate error message

3. Password Encryption

- Given a user registration request, when the user registers, then the system should store the password in an encrypted form.

**Request**
```
POST /api/auth/register
{
  "firstName": String,
  "lastName": String,
  "email": String,
  "password": String,
}
```

**Successful Response**

```
{
 "statusCode": 201
  "message": String,
  "data" : {
        "token": string
        "user": {
          "id": String,
          "first_name": string
          "last_name": string
          "email": String,
          "created_at": Date
        },
   }
}
```

**Error Response**

```
{
  "message": String,
  "error": String,
  "statusCode": Int,
}
```

### **Database Schema**

**Table name: users**

```
Table users {
  id integer [primary key]
  identifier uuid [unique, not null]
  first_name varchar(50) [not null]
  last_name varchar(50) [not null]
  email varchar(255) [unique, not null]
  password varchar(255) [not null]
  created_at timestamp [default: `now()`]
  updated_at timestamp [default: `now()`]

  indexes {
    email
    identifier
  }
}
```

**Database diagram**
https://dbdiagram.io/d/AuthUserSignUp-Db-jideco-6697a1ed8b4bb5230e945ede

### PR DESCRIPTION
**Pull request by Jideco**
# Description
Developed and integrated a seamless user sign up feature to enhance user engagement and onboarding experience. This feature included user registration, account creation, and authentication processes.

**Closes #8 **

# Changes proposed
Accept new user information.

Store in database if not already existed.

Return the appropriate response and status code.

## What were you told to do?
Authentication User Sign Up Endpoint.

## What did you do?
I updated the User table as recommeded [here](https://github.com/hngprojects/hng_boilerplate_csharp_web/issues/8)

- I added UserSignUpDto, UserResponseDto and SignupResponseDto .cs .

- The sign up POST is implemented in AuthController.cs inside Controller folder.

- I made firstname, lastname, email and password field compulsory.

- The email is checked in the Database if it has already existed.

- A token is generated for each user via JWT.

- A response is generated after sign up for further use.

# Check List (Check all the applicable boxes)

- [x]  Email and password, firstname and lastname fields should be required.

- [x]  Password validation should include a minimum of 8 alphanumeric characters.

- [x]  Password should be hashed before being stored in the database.

- [x]  A 422 error response code is returned on failed validations.

- [x]  201 status code is returned after successful signup.

- [x]  JWT token should be generated for the user upon successful sign up and returned as part of the response payload.

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos
Screenshot (317)

### CODE REVIEW

**Comment by Jideco** [comment]
@Godhanded  kudos

**Review by Godhanded — APPROVED** [review]

File: src/Hng.Application/Features/UserManagement/Dtos/SignUpResponse.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿namespace Hng.Application.Features.UserManagement.Dtos
+{
+    public class SignUpResponse
+    {
+        public string Message { get; set; }
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/UserManagement/Dtos/SignUpResponse.cs:5]: Use `JsonPropertyName` attribute to make all your responses snake_case

File: src/Hng.Application/Features/UserManagement/Dtos/UserResponseDto.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Hng.Application.Features.UserManagement.Dtos
+{
+    public class UserResponseDto
+    {
+        public string Id { get; set; }
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/UserManagement/Dtos/UserResponseDto.cs:5]: Same here

File: src/Hng.Web/Controllers/AuthController.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+
+        [HttpPost("register")]
+        [ProducesResponseType(typeof(SignUpResponse), StatusCodes.Status201Created)]
+        [ProducesResponseType(typeof(SignUpResponse), StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(typeof(SignUpResponse), StatusCodes.Status422UnprocessableEntity)]
+        [ProducesResponseType(typeof(SignUpResponse), StatusCodes.Status500InternalServerError)]
+        public async Task<ActionResult<SignUpResponse>> UserSignUp([FromBody] UserSignUpDto body)
+        {
+            if (!ModelState.IsValid)
```
Comment (NdubuisiJr) [inline src/Hng.Web/Controllers/AuthController.cs:27]: No need for this. The Data annotation attribute takes care of the validation

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]
Fix all comments below and add Unit tests

**Comment by Jideco** [comment]
@NdubuisiJr

File: src/Hng.Web/Controllers/AuthController.cs (hunk 3 above)
Comment (Jideco) [reply src/Hng.Web/Controllers/AuthController.cs]: /register  moved to AuthenticationController.cs

File: src/Hng.Application/Hng.Application.csproj [hunk 4]
```diff
@@ -27,7 +27,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Folder Include="Features\UserManagement\Commands\" />
```
Comment (Aniyikaye-Alabi) [inline src/Hng.Application/Hng.Application.csproj:30]: Please do get latest from dev to prevent removing folder paths in the item group.

**Review by Aniyikaye-Alabi — DISMISSED** [review]

File: src/Hng.Application/Hng.Application.csproj (hunk 4 above)
Comment (Jideco) [reply src/Hng.Application/Hng.Application.csproj:30]: well, i pulled from Dev before pushing.

Thanks tho

**Review by NdubuisiJr — APPROVED** [review]

**Pull request merged**